File: frontend/src/app/protocol.ts
export interface GameSnapshot {
  currentPlayer: string;
  nextCardId: number;
  players: PlayerSnapshot[];
}

export interface PlayerSnapshot {
  playerId: string;
  name: string;
  schicksalsmacht: number;
  hand: CardSnapshot[];
  field: CardSnapshot[];
  graveyard: CardSnapshot[];
  deck: CardSnapshot[];
}

export interface CardSnapshot {
  cardId: number;
  owner: string;
  name: string;
  cost: string;
  cardType: string;
  baseStrength: number | null;
  currentStrength: number | null;
  tags: string[];
  description: string;
  modifications: ModificationSnapshot[];
}

export interface ModificationSnapshot {
  kind: string;
  duration: string;
  amount: number | null;
  description: string;
}

export type Prompt =
  | { tag: 'CommandPrompt'; message: string }
  | { tag: 'ChoicePrompt'; message: string; options: string[] }
  | { tag: 'NumberPrompt'; message: string };

export type ServerMessage =
  | { type: 'Connected'; message: string }
  | { type: 'StateSnapshot'; state: GameSnapshot }
  | { type: 'PromptMessage'; prompt: Prompt }
  | { type: 'NoticeMessage'; message: string }
  | { type: 'ErrorMessage'; message: string };

export type ClientMessage =
  | { type: 'SubmitCommand'; command: string }
  | { type: 'SubmitChoice'; choiceIndex: number }
  | { type: 'SubmitNumber'; number: number };

File: src/Web/Protocol.hs
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}

module Web.Protocol where

import Data.Aeson (FromJSON (..), ToJSON (..), defaultOptions, genericParseJSON, genericToEncoding, genericToJSON)
import Data.Aeson.Types (Options (..), SumEncoding (..))
import Data.Text (Text)
import GHC.Generics (Generic)

data ClientMessage
  = SubmitCommand {command :: Text}
  | SubmitChoice {choiceIndex :: Int}
  | SubmitNumber {number :: Int}
  deriving (Eq, Show, Generic)

instance FromJSON ClientMessage where
  parseJSON = genericParseJSON jsonOptions

instance ToJSON ClientMessage where
  toJSON = genericToJSON jsonOptions
  toEncoding = genericToEncoding jsonOptions

data ServerMessage
  = Connected {message :: Text}
  | StateSnapshot {state :: GameSnapshot}
  | PromptMessage {prompt :: Prompt}
  | NoticeMessage {message :: Text}
  | ErrorMessage {message :: Text}
  deriving (Eq, Show, Generic)

instance ToJSON ServerMessage where
  toJSON = genericToJSON jsonOptions
  toEncoding = genericToEncoding jsonOptions

data Prompt
  = CommandPrompt {message :: Text}
  | ChoicePrompt {message :: Text, options :: [Text]}
  | NumberPrompt {message :: Text}
  deriving (Eq, Show, Generic, ToJSON)

data GameSnapshot = GameSnapshot
  { currentPlayer :: Text
  , nextCardId :: Int
  , players :: [PlayerSnapshot]
  }
  deriving (Eq, Show, Generic, ToJSON)

data PlayerSnapshot = PlayerSnapshot
  { playerId :: Text
  , name :: Text
  , schicksalsmacht :: Int
  , hand :: [CardSnapshot]
  , field :: [CardSnapshot]
  , graveyard :: [CardSnapshot]
  , deck :: [CardSnapshot]
  }
  deriving (Eq, Show, Generic, ToJSON)

data CardSnapshot = CardSnapshot
  { cardId :: Int
  , owner :: Text
  , name :: Text
  , cost :: Text
  , cardType :: Text
  , baseStrength :: Maybe Int
  , currentStrength :: Maybe Int
  , tags :: [Text]
  , description :: Text
  , modifications :: [ModificationSnapshot]
  }
  deriving (Eq, Show, Generic, ToJSON)

data ModificationSnapshot = ModificationSnapshot
  { kind :: Text
  , duration :: Text
  , amount :: Maybe Int
  , description :: Text
  }
  deriving (Eq, Show, Generic, ToJSON)

jsonOptions :: Options
jsonOptions =
  defaultOptions
    { sumEncoding = TaggedObject "type" "payload"
    }
